### ISSUE TITLE
I18n configuration in next.config.js breaks app directory

### ISSUE THREAD

**Issue by fabio-nettis** [vercel/next.js#53724]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:56 PDT 2023; root:xnu-8796.141.3~6/RELEASE_X86_64
    Binaries:
      Node: 20.5.0
      npm: 9.8.0
      Yarn: 3.5.0
      pnpm: 8.6.12
    Relevant Packages:
      next: 13.4.14-canary.0
      eslint-config-next: 13.4.13
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router, Internationalization (i18n), Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/fabio-nettis/next-i18n-error

### To Reproduce

Clone the repository and try to navigate to any of the defined pages, you will encounter a not found error for each of the defined routes. This issue is directly related to my previous issue #53665 You can find more about the discover of this bug in [this discussion](https://github.com/vercel/next.js/discussions/53667). 

### Describe the Bug

Since version 13.4.13-canary.0 an application previously working fine on version 13.4.12 returns a 404 status code for every page in the app folder when the `i18n` setting is set in the next config file.

### Expected Behavior

Setting the `i18n` setting is set in the next config file should not break the whole application.

### Which browser are you using? (if relevant)

Chrome 114.0.5735.106 (Official Build) (x86_64)

### How are you deploying your application? (if relevant)

next dev

**Comment by fabio-nettis** [comment]
For anybody that needs a solution now, [here](https://github.com/fabio-nettis/nextjs13-appdir-i18n) is a repository with a working and scalable i18n implementation.

**Comment by vesurbag** [comment]
> For anybody that needs a solution now, [here](https://github.com/fabio-nettis/nextjs13-appdir-i18n) is a repository with a working and scalable i18n implementation that does not use the i18n settings inside `next.config.js`.

Thanks, but it doesn't work in case of default locale shouldn't show in the url, e.g.:
en (default) -> /blog/post
de -> /de/blog/post

**Comment by fabio-nettis** [comment]
@vesurbag Have you tried the [optional folder](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes#optional-catch-all-segments) syntax? Don't know if it only applies to catch-all segments tho.

**Comment by igorovic** [comment]
same issue. tested with next@13.4.20-canary.13

**Comment by dmgawel** [comment]
The issue persists in v13.5.2.

As we have a lot of translated pages in Pages Router, this bug stops us from upgrading Next and adopting the App Router incrementally.

The moment I remove `i18n` config, the App Router works again, but then everything in Pages Router stops working.

**Comment by boredland** [comment]
> The moment I remove i18n config, the App Router works again, but then everything in Pages Router stops working.

And does the app dir pages work in production? For me they only work in development.

**Comment by dmgawel** [comment]
> And does the app dir pages work in production? For me they only work in development.

For us, they do, but we're stuck on v13.2.

**Comment by fabio-nettis** [comment]
> same issue. tested with next@13.4.20-canary.13

Check out the example [repository](https://github.com/fabio-nettis/nextjs13-appdir-i18n) I have provided, it eliminates the need for the `i18n` key inside next.config.js. We currently use it in production and it works just fine.

**Comment by dmgawel** [comment]
> Check out the example [repository](https://github.com/fabio-nettis/nextjs13-appdir-i18n) I have provided, it eliminates the need for the i18n key inside next.config.js. We currently use it in production and it works just fine.

@fabio-nettis Appreciate trying to find a workaround. However, `i18n` config key is not needed for internationalization in App Router, it's meant to be used in the Pages Router. In fact, your example repository implements a standard approach suggested by Next.js docs: https://nextjs.org/docs/app/building-your-application/routing/internationalization

It's not a problem, of course ;-) However, that's not a workaround for this issue.

---

To sum up: Existence of `i18n` config key needed for internationalization in Pages Router breaks App Router, and there is currently no workaround that would enable one to use internationalization in Pages Router and App Router (in general) at the same time.

**Comment by fabio-nettis** [comment]
> > Check out the example [repository](https://github.com/fabio-nettis/nextjs13-appdir-i18n) I have provided, it eliminates the need for the i18n key inside next.config.js. We currently use it in production and it works just fine.
> 
> @fabio-nettis Appreciate trying to find a workaround. However, `i18n` config key is not needed for internationalization in App Router, it's meant to be used in the Pages Router. In fact, your example repository implements a standard approach suggested by Next.js docs: https://nextjs.org/docs/app/building-your-application/routing/internationalization
> 
> It's not a problem, of course ;-) However, that's not a workaround for this issue.
> 
> To sum up: Existence of `i18n` config key needed for internationalization in Pages Router breaks App Router, and there is currently no workaround that would enable one to use internationalization in Pages Router and App Router (in general) at the same time.

@dmgawel Thanks for the response! it would probably be wise to explicitly mention the required absence of the `i18n` key in next.config.js for the app router, since this did not resonate to me personally, and based on the issue activity, suggestively neither did to other people.

Furthermore I'd politely recommend to extends the current documentation on localization inside the app router to be a bit closer to a real-world approach as seen in the repository I provided.

**Comment by mataspetrikas** [comment]
so to summarize the issue - the incremental migration to the app router is not possible for localized NextJs apps, right?

**Comment by dmgawel** [comment]
>  so to summarize the issue - the incremental migration to the app router is not possible for localized NextJs apps, right?

Right 👍

**Comment by mataspetrikas** [comment]
it looks that the `locale` is not being updated in getServerSideProps/getStaticProps from the `next` version `13.4.13`, at least what [this issue](https://github.com/vercel/next.js/issues/53967) seems to suggest:

**Comment by andre-lergier** [comment]
> @dmgawel 
so to summarize the issue - the incremental migration to the app router is not possible for localized NextJs apps, right?

We are on `next` `13.5.4` and I can confirm this conclusion. Does anyone know a version of `next` where it was still working to use App and Pages router simultaneously with i18n?

**Comment by noreiller** [comment]
> > @dmgawel
> > so to summarize the issue - the incremental migration to the app router is not possible for localized NextJs apps, right?
> 
> We are on `next` `13.5.4` and I can confirm this conclusion. Does anyone know a version of `next` where it was still working to use App and Pages router simultaneously with i18n?

On my side, I'm re-implementing the `i18n` feature with a middleware. The only downside is that I cannot remove nor disable the `locale` property on the `next/router` and `next/link` modules. Otherwise, even if I have to update all the links to include the locale in the `href`, it's not so complicated.
This way, the `locale` (from the route segment `[locale]`) is retrieved from the query params and it works like in App Router.

**Comment by guillaumeLamanda** [comment]
It seems to be resolved on the latest release. However, the ##46622 still persist

**Comment by dmgawel** [comment]
> It seems to be resolved on the latest release

Tested on `v13.5.6`, unfortunately it still doesn't work.

**Comment by Abd2rahman** [comment]
It still doesn't work on v14. Has anyone found a workaround to resolve this issue?

**Comment by fabio-nettis** [comment]
> For anybody that needs a solution now, [here](https://github.com/fabio-nettis/nextjs13-appdir-i18n) is a repository with a working and scalable i18n implementation that does not use the i18n settings inside `next.config.js`.

@Abd2rahman Check the linked repository. The only thing this example does not support are optional locale paths.

**Comment by dmgawel** [comment]
I created a repository that recreates the issue with minimal amount of code. It features both Pages and App router example, to highlight the issue when issue both:

https://github.com/dmgawel/next-i18n-issue

**Comment by shinwonse** [comment]
> I created a repository that recreates the issue with a minimal amount of code. It features both Pages and App router example, to highlight the issue when both are used:
> 
> https://github.com/dmgawel/next-i18n-issue

It is same situation with mine🥲 Have you ever tried use `middleware` as Next.js docs said?

**Comment by shinwonse** [comment]
I think I almost find solution from https://github.com/amannn/next-intl/tree/main/examples/example-app-router-migration.

**Comment by Gregorein** [comment]
bump on the thread - what's the status of this issue?
I followed the next docs for the middleware code and I'm hitting the exact same issue somewhere between sending a response and next trying to unravel the [lang] path param

**Comment by Gregorein** [comment]
@shinwonse "almost?"

**Comment by shinwonse** [comment]
> @shinwonse "almost?"

I thought I almost solved it with `next-intl` package, but it was failed and I am still struggling with it 😂

**Comment by Gregorein** [comment]
I got it working using their official doc https://next-intl-docs.vercel.app/docs/getting-started/app-router. I've checked your repo, @shinwonse and it looks like you're almost there, you'd have to use the lang/locale param in the layout instead of using a hook and call the `unstable_setRequestLocale()` with that locale.

**Comment by Multiply** [comment]
We're using a middleware to match domain names and paths, and then we either prepend to, or change the prefix of the path to the fully qualified locale.

This works well for the pages router, but as long as i18n is configured, it won't match `app/[locale]/example/page.tsx`, only `app/example/page.tsx` when visiting `/example`.

The middleware does indeed rewrite to `/en-GB/example`, but the app router seems to remove that part or just ignore the rewrite, when i18n is configured.

**Comment by Alenriquez96** [comment]
Any info on how to solve this problem with output:export? It works fine in local but not when I generate the static files.

**Comment by adil-hublo** [comment]
@dmgawel the way next-intl imports JSON locales in the `i18n.ts` file does not accept anything more than locale as a parameter; which prevent use of namespaces, like in `next-i18next` though

**Comment by adil-hublo** [comment]
@dmgawel sure, this is exactly what I did for the pages router, but for the migrated pages to App Router, I am supposed to use `next-intl` right? And this is where I lack of this "filter" to get only the namespaces I need, instead of all my translation files

**Comment by VMLuca** [comment]
Anything new on this topic? It's a pity that you read everywhere that you can run the app router and the page router in parallel and then it doesn't work because of the internationalisation (really without a workaround).

**Comment by rachellcarroll** [comment]
@dmgawel I tried out next-intl as well, following their app router docs and referencing [next-intl's sample app router migration repo](https://github.com/amannn/next-intl/tree/main/examples/example-app-router-migration). It seems to be working, but I'm getting errors at the console. `Error handling upgrade request TypeError: Cannot read properties of undefined (reading 'bind')
    at DevServer.handleRequestImpl (/Users/<repo_name>/node_modules/next/dist/server/base-server.js:478:50)`

`    "next": "14.2",
    "next-intl": "3.11.3",`
Did you see this too?

**Comment by rachellcarroll** [comment]
@dmgawel I just downgraded to next@13.4.12 and those errors went away!

**Comment by vordgi** [comment]
For those who wanna keep the principle of i18n and static generation but don't want to use [_the excellent for their tasks_] next-intl, I have created the package [@nimpl/router](https://npmjs.com/package/@nimpl/router), which transfers the logic of `rewrites`, `redirects`, `i18n`, and `basePath` from **next.config.js** to **middleware** (_but with the difference that it works_).

Example of a localized application on App Router - https://github.com/vordgi/nimpl-router/blob/main/examples/base/src/middleware.ts;

Documentation - https://nimpl.tech/router

**Comment by JCs3lma** [comment]
Still happening in 14.0.2 version. 🥲

**Comment by joaogarin** [comment]
> Anything new on this topic? It's a pity that you read everywhere that you can run the app router and the page router in parallel and then it doesn't work because of the internationalisation (really without a workaround).

> To sum up: Existence of i18n config key needed for internationalization in Pages Router breaks App Router, and there is currently no workaround that would enable one to use internationalization in Pages Router and App Router (in general) at the same time.

I have to say I agree, this is really depressing after years of app directory being so heavily pushed there is still not path for migrating multilingual apps. Some suggestions here with doing a "manual" middleware implementation I think leave behind the fact that you want your `useRouter()`to have locale info? and your `Link` components in pages to have proper multilingual support (for progressive migration at least) and so on..

Its already what it is to have to figure out all the new patterns in app directory if you still have to do a bunch of changes on all the pages (which are working fine) it becomes a crazy effort to try and do this migration..We have hundreds of components some of which ofc rely on stuff documented in https://nextjs.org/docs/pages/building-your-application/routing/internationalization

https://github.com/vercel/next.js/issues/53724#issuecomment-1994475042 maybe this is a path to explore, still quite sad that we end up in a situation like this after this https://www.youtube.com/watch?v=1-NzQ9ObsfM which was not even 4 years ago

**Comment by Abuzeid1** [comment]
# A possible work around

This solution uses a middleware that removes ```locale``` from ```url```  when its is not default, and sets ```NEXT_LOCALE``` cookie with ```locale``` value. then use ```NEXT_LOCALE``` to get the current language inside ```app``` router.

## Cons
- you need to change how your translation works inside the ```app```
- ```locale``` no longer shows on the url
- you need to change the ```NEXT_LOCALE``` whenever you change ```locale``` inside ```app```

## Pros 
- no code changes required inside ```pages``` 
- changing the language inside ```pages``` changes it inside the ```app```

# Code Example

```ts 
// middlewar.ts
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req: NextRequest) {
  if (req.nextUrl.locale !== req.nextUrl.defaultLocale) {
    const response = NextResponse.redirect(
      req.nextUrl.href.replace(`/${req.nextUrl.locale}`, ""),
    );
    response.cookies.set("NEXT_LOCALE", req.nextUrl.locale);
    return response;
  }
}

export const config = {
  matcher: "/migratedPage",
};
```

Custom translation function which translates based on NEXT_LOCALE cookie value

```ts 
// function.ts
import { cookies } from "next/headers";
const glob = require("glob");

const entries: [string] = glob.sync("locales/ar/*.json");
// type is wrong for typescript resasons
type translate = { [key: string]: translate };

let dict: { ar: translate; en: translate } = { ar: {}, en: {} };
entries.map(async (el) => {
  const str = el.toString().replace(".json", "").replace("locales/ar/", "");
  dict["ar"][str] = await import(`locales/ar/${str}`);
  dict["en"][str] = await import(`locales/en/${str}`);
});

;

export const customT = (str: string): string => {
  // convection filname:key add .subKey if there is one
  let arr = str.split(":");
  const locale: "ar" | "en" =
    cookies().get("NEXT_LOCALE")?.value === "ar" ? "ar" : "en";

  let tr: translate | string = dict[locale];
  tr = tr[arr[0]];

  arr = arr[1].split(".");

  while (arr.length > 0) {
    tr = tr[arr[0]] as translate;
    arr.shift();
  }
  // @ts-ignore onlo to keep typescript happy
  return tr as string;
};
```

**Comment by kamalendugarai** [comment]
Nextjs15 has been introduced!! Is there any legitimate solution to this issue available. Or we all are missing something here, please suggest.

**Comment by baba43** [comment]
So this is still not working? Then please remove the i18n as a feature from your docs and also remove the Wiki page.

**Comment by LeulAria** [comment]
any update on this any one ?

**Comment by Frontend-geonu** [comment]
I'm experiencing the same issue in next@15.3.3 😞

**Comment by adil-hublo** [comment]
The pain to migrate in a project reliably using i18n is so hard I am asking myself if it's not easier to directly switch to Tanstack Router 🤦‍♂️

**Comment by mehdirussel** [comment]
no update in next 16... just a heads up for all of you I suggest not to use app router until it's fully stable

**Comment by RobVermeer** [comment]
This appears to be the same root cause as #86048: the Pages Router `i18n` config strips the locale prefix before routing, which breaks App Router routes that carry the locale as a segment (e.g. `app/[lang]/...`).

I've opened a fix in #94606 — it covers dev, production/`next start`, and static export, with e2e tests. If anyone here can test against that branch, feedback would be welcome.
